hazelcast/src/main/java/com/hazelcast/instance/impl/ClusterTopologyIntentTracker.java

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (c) 2008-2022, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.instance.impl;
+
+/**
+ * Receives updates about the context in which Hazelcast is executed
+ * in order to detect what is the intent of topology changes from the
+ * managed runtime context that may affect how the Hazelcast cluster should react.
+ * <p/>
+ * Terminology:
+ * <ul>
+ *     <li><b>Managed runtime context</b>: a runtime orchestration environment within which Hazelcast is
+ *     being executed (e.g. Kubernetes).</li>
+ *     <li><b>Replica</b>: a Hazelcast member that is executed in a managed runtime context. For example,
+ *     when running a 3-member Hazelcast cluster, there are 3 replicas.</li>
+ *     <li><b>Specification / specified replicas</b>: the desired runtime state of the Hazelcast member, as declared by the user
+ *     to the managed runtime context. For example, when user executes {@code kubectl scale sts hz --replicas 5},
+ *     the <b>specified replicas</b> are 5.</li>
+ *     <li><b>Current replicas</b>: number of replicas that are running by the managed runtime. Notice that current replicas
+ *     does not necessarily reflect the current number of members in the Hazelcast cluster. Examples:
+ *     <br/>- When a kubernetes pod is terminated, "current replicas" is immediately
+ *     decreased by 1, even though the Hazelcast member is still live and just started performing its graceful shutdown. So
+ *     there will be some time during which current replicas may be lower than actual number of members in the Hazelcast cluster
+ *     (until graceful shutdown completes).
+ *     <br/>- When a kubernetes pod is scheduled, "current replicas" immediately increases by 1 even though Hazelcast is just
+ *     started and not joined to any cluster.</li>
```

Comment:
I think we can mention what happens to currentReplicas during stateful set restart here. As we don't sum it with `updateReplicas` now, it can go to 0, while `readyReplicas` fluctuate between `specSize` and `specSize - 1`.